### DISCUSSION TITLE
Do we need MethodRouter::route_layer?

### DISCUSSION THREAD

**Discussion by zishon** [tokio-rs/axum#2136] [Q&A] [answered]
I am learning middleware of axum, as MethodRouter::layer is always bind to a specific method_router.

And the document says: 

This works similarly to [MethodRouter::layer](https://docs.rs/axum/0.6.19/axum/routing/method_routing/struct.MethodRouter.html#method.layer) except the middleware will only run if the request matches a route.

So I wonder in what case rather using  MethodRouter::route_layer than MethodRouter::layer?

**Comment by davidpdrsn** [comment, marked as answer]
We do. Middleware’s added with `layer` also run for the fallback. If you don’t want that you can use `route_layer`.

Same applies for `Router`

**Reply by zishon** [reply]
Now I get the difference of this two, thanks.

**Reply by zishon** [reply]
Is it necessary to update the document here [MethodRouter::route_layer](https://docs.rs/axum/0.6.19/axum/routing/method_routing/struct.MethodRouter.html#method.route_layer) to make the difference between layer and route_layer more pronounced?

**Reply by davidpdrsn** [reply]
I don’t think so but you’re welcome to submit a PR.
